R/case-variables.R

```diff
@@ -0,0 +1,169 @@
+        halt("A case must be a list")
+    }
+    wrong_case_names <- setdiff(names(case), c("id", "name", "expression",
+                                               "numeric_value", "missing"))
+    if (length(wrong_case_names) > 0) {
+        halt("each case must have at most an id, name, expression, ",
+             "numeric_value, and missing element. The errant arguments were: ",
+             serialPaste(wrong_case_names))
+    }
+    if (!is.null(case$id) && !is.whole(case$id)) {
+        halt("a case's id must be an integer")
+    }
+    if (!is.character(case$name)) {
+        halt("a case's name must be a character")
+    }
+    if (class(case$expression) != "CrunchLogicalExpr") {
+        halt("a case's expression must be a CrunchLogicalExpr")
+    }
+    if (is.null(case$numeric_value)) {
+        case['numeric_value'] <- list(NULL)
+    } else if (!is.numeric(case$numeric_value)) {
+        halt("a case's numeric_value must be a numeric")
+    }
+    if (is.null(case$missing)) {
+        case['missing'] <- FALSE
+    } else if (!is.logical(case$missing)) {
+        halt("a case's missing must be a logical")
+    }
+
+    return(case)
+}
+
+## Check for "integers" but don't require true integer storage
+is.whole <- function (x) is.numeric(x) && floor(x) == x
+
+ensureValidCases <- function(cases, else_case) {
+    cases <- lapply(cases, ensureValidCase)
+
+    if (!missing(else_case)) {
+        # assign API default values if not given
+        defaults <- list(id=NULL, name=NULL, numeric_value=NULL, missing=FALSE)
+        else_case <- modifyList(defaults, else_case)
+
+        if(!is.null(else_case$expression)) {
+            halt("else_cases should not have any conditions expression")
+        }
+        if(!is.character(else_case$name)) {
+            halt("else_cases must have a (character) name")
+        }
+        if(!is.null(else_case$id) & !is.whole(else_case$id)) {
+            halt("id must be an integer")
+        }
+        cases <- c(cases, list(else_case))
+    }
+
+    # make ids if the cases don't have ids.
+    need_ids <- vapply(cases, function (x) is.null(x$id), logical(1))
+    supplied_ids <- vapply(cases[!need_ids], vget("id"), numeric(1))
+    new_ids <- seq_along(cases) # generate too many ids
+    new_ids <- setdiff(new_ids, supplied_ids) # discard ids where there is overlap
+    cases[need_ids] <- lapply(seq_along(cases[need_ids]), function(i) {
+        cases[need_ids][[i]]$id <- new_ids[i]
+        return(cases[need_ids][[i]])
+    })
+
+    # check there are no duplicate ids and they are less than 2^15-1
+    all_ids <- vapply(cases, vget("id"), numeric(1))
+    if (anyDuplicated(all_ids) > 0) {
+        halt("there are duplicate ids provided: ", serialPaste(all_ids))
+    }
+    if (any(all_ids > 32767L)) {
+        halt("id must be less than 32,768, this might be a result of too many cases being used.")
+    }
+    if (any(all_ids < 1L)) {
+        halt("id must not be less than 1")
+    }
+
```

Comment:
I've fixed both of these, as well as adding a check in case someone tries to supply multiple cases in a single list (eg `cases=list(name=c("A", "B"), expression=c(ds$v1 == "A", ds$v2 == "B"))`, unlikely, but the error that is produced when that is done isn't super transparent.). I've also added a check if any of the expressions are duplicates. This wouldn't *break* anything since whichever is first will be matched will be used, I could imagine someone accidentally doing this and being confused by the result.